### ISSUE TITLE
OpenCL Program cache not used consistently; some paths skip cache and always go to disk

### ISSUE THREAD

**Issue by diablodale** [opencv/opencv#19570]
The in-memory `ocl::Program` cache `ocl::Context::Impl::phash` is not used consistently across the various methods to create/get/compile an `ocl::Program`. Result? Codepaths without this cache seek and load files from the harddrive on every kernel create -- strongly undesired. As example, I run a custom opencl kernel. This errant cache behavior adds 2 milliseconds to total kernel execution time for every frame. 😢

#### System information (version)
- OpenCV => 4.5.0+
- Operating System / Platform => all
- Compiler => all

#### Detailed description

The errant codepath is exposed in the following pseudocode. The below will always go to disk. It never checks the in-memory cache.

```c++
ocl::ProgramSource prog_source{module, name, sourcecode, ""};
ocl::Program prog(prog_source, build_flags, errmsg); // no in-memory cache; always goes to disk
```

Part of the cause is the location of this cache management code. Currently it is within `Context::Impl`. I believe it should instead be within `Program::Impl` near `Program::Impl::compileWithCache()`. The function `compileWithCache()` already contains an on-disk cache. I believe centralizing `Program` cache code in `Program::Impl` would be better. OpenCV would then have a centralized L1 (in-memory) and L2 (disk) cache for all `Program`. 🤩

Some opencv codepaths do use the in-memory cache. For example, if use the single API that compiles and creates a Kernel in one step, the in-memory cache *is* used -- its codepath calls `Context::getProg()` which currently contains the in-memory cache code.

```c++
ocl::Kernel kern(name, prog_source, build_flags, errmsg); // this uses in-memory cache via Context::getProg()
```

I do not think it is difficult to resolve this issue. The existing L1 (in-memory) cache code is easy to understand and in a single place in code. Most of it can be shifted to `Program::Impl` with few changes. And having it there will enable all codepaths that create/get/compile an `ocl::Program` to gain the advantage of both L1 (in-memory) and L2 (disk) cache.

Comments? 🧐

##### Issue submission checklist

 - [X] I report the issue, it's not a question
 - [X] I checked the problem with documentation, FAQ, open issues,
 - [X] I updated to latest OpenCV version and the issue is still there
 - [X] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by alalek** [comment]
Is your case similar to [this test code](https://github.com/opencv/opencv/blob/4.5.1/modules/core/test/ocl/test_opencl.cpp#L76)?

**Comment by diablodale** [comment]
No. It is the first code in the section "Detailed description". When ProgramSource and Program are created from text source code (not a binary), the codepath doesn't use the L1 cache.

**Comment by alalek** [comment]
or [this](https://github.com/opencv/opencv/blob/4.5.1/modules/core/test/ocl/test_opencl.cpp#L62)?

**Comment by diablodale** [comment]
Yes. When a `ProgramSource` is created from text, and then a `Program` created from that, that codepath never calls `Context::getProg()`. The *only* function that implements the L1 (in-memory) cache is `getProg()`.
Therefore this text source code -> program codepath always goes to disk.

The opencv built-in ocl kernels/functions usually use `ocl::Kernel kern(name, prog_source, build_flags, errmsg)`. That one-line approach *does* call `getProg()` and therefore uses the L1 cache.
